add doc on experiment rules syntax (#1355)
semgrep/semgrep-docs: docs/writing-rules/experiments/pattern-syntax.md

---
slug: pattern-syntax
title: Pattern syntax (Experimental)
hide_title: true
description: Learn how to use Semgrep's experimental pattern syntax to search code for a specific code pattern.
tags:
  - Rules
  - Semgrep Code
---

## Pattern syntax (experimental)

Patterns are the expressions Semgrep uses to match code when it scans for vulnerabilities. This article describes the new syntax for Semgrep pattern operators. See [Pattern syntax](/writing-rules/pattern-syntax) for information on the existing pattern syntax.

There is often a one-to-one translation from the existing syntax to the experimental syntax. These changes are marked with <i class= "fa-solid fa-diamond"></i>. However, some changes are quite different. These changes are marked with <i class="fa-solid fa-exclamation"></i>

:::warning

> These patterns are **experimental** and subject to change.

Reviewer: ```suggestion
* These patterns are **experimental** and subject to change.
```

> 

Reviewer: ```suggestion
```

> You can't mix and match existing pattern syntax with the experimental syntax.

Reviewer: ```suggestion
* You can't mix and match existing pattern syntax with the experimental syntax.
```

Replaces [pattern-either](/writing-rules/rule-syntax/#pattern-either). Matches any of the patterns specified.

```yaml
any:
  - <pat1>
  - <pat2> 
    ...
  - <patn>
```

### <i class="fa-solid fa-diamond"></i> `all`

Replaces [patterns](/writing-rules/rule-syntax/#patterns). Matches all of the patterns specified.

```yaml
all:
  - <pat1>
  - <pat2> 
    ...
  - <patn>
```

### <i class="fa-solid fa-diamond"></i> `inside`

Replaces [pattern-inside](/writing-rules/rule-syntax/#pattern-inside). Match any of the sub-patterns inside of the primary pattern.

```yaml
inside: 
  any: 
    - <pat1> 
    - <pat2>
```

Alternatively:

```yaml
any:
  - inside: <pat1>
  - inside: <pat2>
```

### <i class="fa-solid fa-diamond"></i> `not`

Replaces [pattern-not](/writing-rules/rule-syntax/#pattern-not). Accepts any pattern and does **not** match on those patterns.

```yaml
not:
  any:
    - <pat1>
    - <pat2>
```

Alternatively:

```yaml
all:
  - not: <pat1>
  - not: <pat2>
```

### <i class="fa-solid fa-diamond"></i> `regex`

Replaces [pattern-regex](/writing-rules/rule-syntax/#pattern-regex) Matches based on the regex provided.

```yaml
regex: "(.*)"
```

## Metavariables

> Metavariables are an abstraction to match code when you don't know the value or contents beforehand. They're similar to [capture groups](https://regexone.com/lesson/capturing_groups) in regular expressions and can track values across a specific code scope. This

Reviewer: (This nit is completely optional and can be ignored.) Do we want to standardize the definition of a metavariable?

includes variables, functions, arguments, classes, object methods, imports,
exceptions, and more.

Metavariables begin with a `$` and can only contain uppercase characters, `_`, or digits. Names like `$x` or `$some_value` are invalid. Examples of valid metavariables include `$X`, `$WIDGET`, or `$USERS_2`.

### <i class="fa-solid fa-exclamation"></i> `where`

Unlike Semgrep's existing pattern syntax, the following no longer occur under `pattern` or `all`: `metavariable-pattern`, `metavariable-regex`, `metavariable-comparison`, `metavariable-analysis`, and `focus-metavariable`. These must occur within a `where` clause.

> A `where` clause is a companion to a pattern. It indicates that Semgrep should match based on the pattern if all the conditions hold.

Reviewer: ```suggestion
A `where` clause is a companion to a pattern. It indicates that Semgrep should match based on the pattern if all the conditions are true.
```
Just a suggestion to simplify the English a bit.
